File: packages/luxar/src/luxar/io/_ordering/lines.py
"""Lines spatial ordering (dual: vertices in D-space + segments in 2D-space).

Line vertices are ordered like Points (in D-space); segments are ordered in
(2xD)-space by concatenating both endpoints, so spatially-coherent segments land
in the same chunk. Chunk bounds include the line-width extent on spatial axes.
"""

from __future__ import annotations

from typing import Literal, Optional

import numpy as np

from luxar.core import Dimension

from .bounds import _BARRIER_BOUND_EPS
from .curves.hilbert import hilbert_encode_nd
from .curves.morton import morton_encode_128bit, morton_encode_nd
from .grid import normalize_coords_to_grid
from .points import sort_points_compound


def convert_to_indexed(
    n_vertices: int, line_type: str, indices: Optional[np.ndarray]
) -> np.ndarray:
    """Convert any line type to indexed segment pairs.

    All line types are internally converted to the unified indexed representation
    for efficient spatial ordering and storage.

    Args:
        n_vertices: Number of vertices in the lines
        line_type: One of "segments", "polyline", "loop", "indexed"
        indices: For "indexed" type, the user-provided index array

    Returns:
        segments: (S, 2) uint32 array of vertex index pairs

    Raises:
        ValueError: If line_type is invalid or indices missing for "indexed" type
    """
    if line_type == "indexed":
        if indices is None:
            raise ValueError("Indexed line type requires indices array")
        return indices.reshape(-1, 2).astype(np.uint32)
    elif line_type == "segments":
        # Each pair of consecutive vertices forms a segment
        return np.arange(n_vertices, dtype=np.uint32).reshape(-1, 2)
    elif line_type == "polyline":
        # Connect consecutive vertices: (0,1), (1,2), (2,3), ...
        return np.column_stack(
            [
                np.arange(n_vertices - 1, dtype=np.uint32),
                np.arange(1, n_vertices, dtype=np.uint32),
            ]
        )
    elif line_type == "loop":
        # Like polyline but connect last to first
        return np.column_stack(
            [
                np.arange(n_vertices, dtype=np.uint32),
                np.roll(np.arange(n_vertices, dtype=np.uint32), -1),
            ]
        )
    else:
        raise ValueError(
            f"Invalid line_type '{line_type}'. Must be one of: segments, polyline, loop, indexed"
        )


def sort_segments_compound(
    segment_coords_2d: np.ndarray,
    dimensions: list[Dimension],
    method: Literal["morton", "hilbert"] = "hilbert",
) -> tuple[np.ndarray, dict]:
    """Sort segments using compound ordering in (2×D)-space.

    Segments are represented as (2×D)-dimensional points by concatenating both
    endpoint coordinates. This captures the full geometric nature of segments
    (position, orientation, length) for spatial coherence.

    The compound ordering strategy is:
    - Primary sort: Discrete dimensions from both endpoints (lexicographic)
    - Secondary sort: Morton/Hilbert code of spatial dimensions from both endpoints

    Args:
        segment_coords_2d: Segment coordinates in (2×D)-space, shape (S, 2*d)
            First d columns are start point, second d columns are end point.
        dimensions: Original Dimension objects (may have more dims than data)
        method: Spatial curve method ("morton" or "hilbert"), default "hilbert"

    Returns:
        sort_indices: Indices to reorder segments
        metadata: Dict with ordering metadata in (2×D)-space
    """
    n_segments, n_dims_2d = segment_coords_2d.shape
    # Use actual data dimensionality (segment_coords_2d is 2×D, so D = n_dims_2d // 2)
    n_dims_original = n_dims_2d // 2
    # Use only the first n_dims_original dimensions from the dimensions list
    dimensions = dimensions[:n_dims_original]

    # Build (2×D) dimension classification
    # Discrete dims from both endpoints: if dim 3,4 are discrete in D-space,
    # then dims 3,4,d+3,d+4 are discrete in (2×D)-space
    slice_dims_2d = []
    ordering_dims_2d = []

    for i, d in enumerate(dimensions):
        if d.discrete and not d.display:
            slice_dims_2d.append(i)  # Start point discrete dim
            slice_dims_2d.append(n_dims_original + i)  # End point discrete dim
        else:
            ordering_dims_2d.append(i)  # Start point spatial dim
            ordering_dims_2d.append(n_dims_original + i)  # End point spatial dim

    # Compute bits per ordering dimension (auto-select 128-bit if needed)
    if ordering_dims_2d:
        bits_per_dim = min(21, 64 // len(ordering_dims_2d))
        use_128bit = bits_per_dim < 10
        if use_128bit:
            bits_per_dim = min(21, 128 // len(ordering_dims_2d))
    else:
        bits_per_dim = 21
        use_128bit = False

    # Extract ordering dimension coordinates
    spatial_codes: tuple[np.ndarray, np.ndarray] | np.ndarray
    if ordering_dims_2d:
        ordering_coords = segment_coords_2d[:, ordering_dims_2d]
        ordering_min = ordering_coords.min(axis=0)
        ordering_max = ordering_coords.max(axis=0)

        # Normalize to grid
        grid_coords = normalize_coords_to_grid(
            ordering_coords, ordering_min, ordering_max, 2**bits_per_dim
        )

        # Compute spatial curve codes
        if method == "morton":
            if use_128bit:
                high, low = morton_encode_128bit(grid_coords, bits_per_dim)
                spatial_codes = (high, low)  # Tuple for lexsort
            else:
                spatial_codes = morton_encode_nd(grid_coords, bits_per_dim)
        elif method == "hilbert":
            # Hilbert doesn't have 128-bit implementation - fall back to morton for high dims
            if use_128bit:
                high, low = morton_encode_128bit(grid_coords, bits_per_dim)
                spatial_codes = (high, low)
            else:
                spatial_codes = hilbert_encode_nd(grid_coords, bits_per_dim)
        else:
            raise ValueError(f"Unknown method: {method}")
    else:
        # No ordering dimensions - all discrete
        spatial_codes = np.zeros(n_segments, dtype=np.uint64)
        ordering_min = np.array([])
        ordering_max = np.array([])

    # Create compound sort key
    if slice_dims_2d:
        # Extract discrete dimension values
        slice_values = segment_coords_2d[:, slice_dims_2d]

        # Create sort keys: (discrete_tuple, spatial_code)
        if isinstance(spatial_codes, tuple):
            # 128-bit: lexsort by (low, high, discrete_dims...)
            high, low = spatial_codes
            sort_indices = np.lexsort(
                [low, high]
                + [slice_values[:, i] for i in range(len(slice_dims_2d) - 1, -1, -1)]
            )
        else:
            sort_indices = np.lexsort(
                [spatial_codes]
                + [slice_values[:, i] for i in range(len(slice_dims_2d) - 1, -1, -1)]
            )
    else:
        # Pure spatial ordering (no discrete dimensions)
        if isinstance(spatial_codes, tuple):
            high, low = spatial_codes
            sort_indices = np.lexsort([low, high])
        else:
            sort_indices = np.argsort(spatial_codes)

    # Build metadata
    metadata = {
        "ordering": method,
        "slice_dims": slice_dims_2d,
        "ordering_dims": ordering_dims_2d,
        "ordering_min": ordering_min.tolist() if len(ordering_min) > 0 else [],
        "ordering_max": ordering_max.tolist() if len(ordering_max) > 0 else [],
        "ordering_bits_per_dim": bits_per_dim,
    }

    return sort_indices, metadata


def order_lines_spatial(
    vertices: np.ndarray,
    segments: np.ndarray,
    dimensions: list[Dimension],
    method: Literal["morton", "hilbert"] = "hilbert",
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, dict]:
    """Apply dual spatial ordering to lines using Morton or Hilbert curves.

    Lines have dual ordering:
    1. Vertices are ordered in D-space (like Points)
    2. Segments are ordered in (2×D)-space (concatenating both endpoints)

    This is the main entry point for Lines spatial ordering.

    Args:
        vertices: (V, D) float32 vertex positions
        segments: (S, 2) uint32 index pairs (from convert_to_indexed)
        dimensions: List of Dimension objects (may have more dims than data)
        method: Ordering method ("morton" or "hilbert")

    Returns:
        sorted_vertices: (V, D) float32 reordered vertex positions
        sorted_segments: (S, 2) uint32 reordered and remapped segment indices
        vertex_sort_indices: Indices to recover original vertex order
        segment_sort_indices: Indices to recover original segment order
        metadata: Dict with "vertex_ordering" and "segment_ordering" sub-dicts
    """
    _V, D = vertices.shape  # V unused but kept for reference

    # Use only the first D dimensions from the dimensions list
    # This handles cases where scene has more dimensions than data
    dimensions = dimensions[:D]

    # 1. Order vertices in D-space (with compound ordering for discrete dims)
    vertex_sort_indices, vertex_metadata = sort_points_compound(
        vertices, dimensions, method=method
    )
    sorted_vertices = vertices[vertex_sort_indices]

    # 2. Create inverse mapping for index remapping
    # After sorting, vertex at original index i is now at inverse_map[i]
    inverse_map = np.argsort(vertex_sort_indices).astype(np.uint32)
    remapped_segments = inverse_map[segments]

    # 3. Build (2×D) segment coordinates for ordering
    # Each segment becomes a 2D point: [x1, y1, z1, x2, y2, z2] for 3D
    segment_coords_2d = np.concatenate(
        [
            sorted_vertices[remapped_segments[:, 0]],  # Start points
            sorted_vertices[remapped_segments[:, 1]],  # End points
        ],
        axis=1,
    )  # Shape: (S, 2×D)

    # 4. Order segments in (2×D)-space (with compound ordering)
    segment_sort_indices, segment_metadata = sort_segments_compound(
        segment_coords_2d, dimensions, method=method
    )
    sorted_segments = remapped_segments[segment_sort_indices]

    return (
        sorted_vertices,
        sorted_segments,
        vertex_sort_indices,
        segment_sort_indices,
        {"vertex_ordering": vertex_metadata, "segment_ordering": segment_metadata},
    )


def compute_vertex_chunk_bounds(
    vertices: np.ndarray,
    chunk_size: int,
    slice_dims: Optional[list[int]] = None,
) -> np.ndarray:
    """Compute chunk bounding boxes for vertices (no radius/width expansion).

    Args:
        vertices: Vertex positions (already sorted), shape (V, D)
        chunk_size: Number of vertices per chunk
        slice_dims: Indices of discrete (non-spatial) dimensions (padded by a
            float-boundary epsilon only — the reader's per-dimension tolerance
            owns the query reach)

    Returns:
        chunk_bounds: Bounding boxes, shape (num_chunks, D, 2)
    """
    n_vertices, n_dims = vertices.shape
    num_chunks = (n_vertices + chunk_size - 1) // chunk_size
    discrete_dims = set(slice_dims) if slice_dims else set()

    chunk_bounds = np.zeros((num_chunks, n_dims, 2), dtype=np.float32)

    for chunk_idx in range(num_chunks):
        start_idx = chunk_idx * chunk_size
        end_idx = min(start_idx + chunk_size, n_vertices)
        chunk_verts = vertices[start_idx:end_idx]

        for d in range(n_dims):
            if d in discrete_dims:
                # Discrete: tight bounds padded by a float-boundary epsilon only
                # (the reader's per-dimension tolerance owns the query reach).
                chunk_bounds[chunk_idx, d, 0] = (
                    chunk_verts[:, d].min() - _BARRIER_BOUND_EPS
                )
                chunk_bounds[chunk_idx, d, 1] = (
                    chunk_verts[:, d].max() + _BARRIER_BOUND_EPS
                )
            else:
                # Spatial: exact bounds (no size expansion for vertices)
                chunk_bounds[chunk_idx, d, 0] = chunk_verts[:, d].min()
                chunk_bounds[chunk_idx, d, 1] = chunk_verts[:, d].max()

    return chunk_bounds


def compute_segment_chunk_bounds(
    vertices: np.ndarray,
    segments: np.ndarray,
    widths: np.ndarray,
    chunk_size: int,
    slice_dims: Optional[list[int]] = None,
) -> np.ndarray:
    """Compute chunk bounding boxes for segments (includes line width).

    Segment bounds are in D-dimensional space (not 2×D) for view frustum
    intersection tests. Each segment's bounds include the line width extent.

    IMPORTANT: widths must be a full (V,) array. Broadcast widths should be
    expanded with np.full(V, width_value) before calling this function.

    Args:
        vertices: Vertex positions (already sorted), shape (V, D)
        segments: Segment index pairs (already sorted), shape (S, 2)
        widths: Vertex widths (already sorted), shape (V,)
        chunk_size: Number of segments per chunk
        slice_dims: Indices of discrete (non-spatial) dimensions (padded by a
            float-boundary epsilon only — the reader's per-dimension tolerance
            owns the query reach)

    Returns:
        chunk_bounds: Bounding boxes, shape (num_chunks, D, 2)
    """
    _V, D = vertices.shape
    S = segments.shape[0]
    num_chunks = (S + chunk_size - 1) // chunk_size
    discrete_dims = set(slice_dims) if slice_dims else set()

    chunk_bounds = np.zeros((num_chunks, D, 2), dtype=np.float32)

    for chunk_idx in range(num_chunks):
        start_idx = chunk_idx * chunk_size
        end_idx = min(start_idx + chunk_size, S)
        chunk_segs = segments[start_idx:end_idx]

        # Get vertex positions and widths for this chunk's segments
        p1 = vertices[chunk_segs[:, 0]]  # Start vertex positions
        p2 = vertices[chunk_segs[:, 1]]  # End vertex positions
        w1 = widths[chunk_segs[:, 0]]  # Start vertex widths
        w2 = widths[chunk_segs[:, 1]]  # End vertex widths
        max_w = np.maximum(w1, w2)  # Conservative bound per segment

        for d in range(D):
            if d in discrete_dims:
                # Discrete: no width expansion, only a float-boundary epsilon.
                # The reader's per-dimension tolerance owns the query reach
                # (see _BARRIER_BOUND_EPS).
                chunk_bounds[chunk_idx, d, 0] = (
                    min(p1[:, d].min(), p2[:, d].min()) - _BARRIER_BOUND_EPS
                )
                chunk_bounds[chunk_idx, d, 1] = (
                    max(p1[:, d].max(), p2[:, d].max()) + _BARRIER_BOUND_EPS
                )
            else:
                # Spatial: include width extent
                chunk_bounds[chunk_idx, d, 0] = min(
                    (p1[:, d] - max_w).min(), (p2[:, d] - max_w).min()
                )
                chunk_bounds[chunk_idx, d, 1] = max(
                    (p1[:, d] + max_w).max(), (p2[:, d] + max_w).max()
                )

    return chunk_bounds

File: packages/luxar-viewer/src/types/lines.ts
/**
 * Lines type definitions for luxar-viewer.
 *
 * These types mirror Python luxar.core.Lines and enable type-safe
 * handling of line data throughout the viewer.
 *
 * @module types/lines
 */

import type { BlendingMode } from './blending';
import { LINE_TYPES, type LineTypeName, type OrderingMethodName } from './format-contract';
import type { ViewState } from '../data/data-loader-types';
import type { ScalarArray } from './points';
import type { PositionBounds } from './zarr';
import type { LoaderMetrics, MonitorEventListener, QueryInfo } from './data-monitor-types';

// ============================================================================
// Metadata Types (from zarr .zattrs)
// ============================================================================

/**
 * Ordering metadata structure (shared between vertices and segments)
 *
 * Lines use dual spatial indexing:
 * - Vertices: Ordered in D-dimensional space using space-filling curves
 * - Segments: Ordered in (2×D)-dimensional space (start+end concatenated)
 */
export interface OrderingMetadata {
  /** Discrete dimension indices for compound ordering */
  slice_dims: number[];

  /** Spatial dimension indices for curve ordering */
  ordering_dims: number[];

  /** Min bounds for coordinate normalization */
  ordering_min: number[];

  /** Max bounds for coordinate normalization */
  ordering_max: number[];

  /** Elements per chunk */
  chunk_size: number;

  /** Bits per dimension for space-filling curve encoding */
  ordering_bits_per_dim?: number;
}

/**
 * Original line type from Python API.
 *
 * - 'segments': Independent line segments (pairs of vertices)
 * - 'polyline': Connected vertices forming a continuous line
 * - 'loop': Polyline with first and last vertex connected
 * - 'indexed': Explicit vertex/segment indices (most flexible)
 */
export type LineType = LineTypeName;

/**
 * Lines node metadata from zarr .zattrs
 *
 * This interface defines all metadata stored with a Lines node in the zarr archive.
 * It enables type-safe access to line properties throughout the viewer.
 */
export interface LinesMetadata {
  /** Node type identifier */
  type: 'lines';

  /** Total vertex count */
  n_vertices: number;

  /** Total segment count */
  n_segments: number;

  /** Vertex position dimensionality */
  ndim: number;

  /** Original line type from Python API */
  original_line_type: LineType;

  /** Maximum line width in world units */
  max_width: number;

  /**
   * Authored nD axis-aligned vertex bounds. Written unconditionally on
   * every lines node by the compiler (`geometry_writers/lines.py`), and
   * on an additive-ladder parent as the union over its levels — so it is
   * the one authored extent that survives when a node has no spatial
   * index. Optional only for hand-written metadata.
   */
  position_bounds?: PositionBounds;

  /** Whether colors array is present */
  has_colors: boolean;

  /** Whether sharpness array is present */
  has_sharpness: boolean;

  /**
   * Whether per-vertex scalar values for colormap lookup exist.
   *
   * gates whether the loader opens the `scalars` zarr array and
   * whether `NodeFactory.createLinesNode` enables `USE_COLORMAP`.
   */
  has_scalars?: boolean;

  /**
   * Named colormap or 'custom' (paired with `colormap_lut` zarr array).
   * viewer reads this to choose the LUT texture; when set without
   * `has_scalars`, the colormap path is suppressed (fail-closed).
   */
  colormap?: string;

  /**
   * `[min, max]` for normalising scalars before LUT lookup.
   * passed to `material.updateScalarRange` so a value of `min`
   * samples LUT index 0 and `max` samples 255.
   */
  scalar_data_range?: [number, number];

  /** Whether spatial index exists (redundant with ordering !== 'none', but explicit) */
  has_spatial_index?: boolean;

  /** Whether per-element string labels exist (CSR-encoded, for hover tooltips) */
  has_labels?: boolean;

  /** Whether per-element image labels exist (CSR-encoded, for hover thumbnails) */
  has_image_labels?: boolean;

  /** Whether per-element stable string keys exist (CSR-encoded, for element actions) */
  has_keys?: boolean;

  /** Spatial ordering method */
  ordering: OrderingMethodName;

  /** Vertex ordering metadata (when ordering != 'none') */
  vertex_ordering?: OrderingMetadata;

  /** Segment ordering metadata (when ordering != 'none') */
  segment_ordering?: OrderingMetadata;

  /** 4x4 transform matrix (column-major for THREE.js) */
  transform?: number[];

  /** Opacity multiplier */
  opacity?: number;

  /** Absorption coefficient κ (volumetric blending mode; identity 1.0) */
  absorption?: number;

  /** Gamma correction */
  gamma?: number;

  /** Intensity (linear color multiplier / gain) */
  intensity?: number;

  /** Offset (additive brightness shift / black level subtraction) */
  offset?: number;

  /** Blending mode */
  blending_mode?: BlendingMode;

  /** Whether this node is exposed as a layer in the Layers panel */
  layer?: boolean;

  /**
   * List of dimension names to extend visibility across.
   * Lines with extend_to_all will be visible regardless of slice position
   * in the specified dimensions (e.g., ['time'] makes geometry visible
   * at all time values).
   */
  extend_to_all?: string[];
}

// ============================================================================
// Loaded Data Types
// ============================================================================

/**
 * Segment range for partial loading (mirrors PointRange pattern)
 */
export interface SegmentRange {
  /** Start segment index (inclusive) */
  start: number;

  /** End segment index (exclusive) */
  end: number;
}

/**
 * Raw lines data loaded from zarr before nD projection.
 *
 * At this stage:
 * - Vertices are in full nD space
 * - Segments use LOCAL indices into the loaded vertex arrays
 * - Attributes may need broadcasting (scalar → per-vertex)
 */
export interface LoadedLinesData {
  /** Vertex positions (N vertices * ndim dimensions), flattened row-major */
  positions: Float32Array;

  /** Segment index pairs (M segments * 2) - local indices into positions array */
  segments: Uint32Array;

  /** Vertex widths (N,) or (1,) if broadcast */
  widths: Float32Array;

  /** Vertex colors (N * colorComponents) RGB or RGBA, null if not present
   * Supports Float32Array (HDR), Uint8Array (SDR), or Uint16Array */
  colors: Float32Array | Uint8Array | Uint16Array | null;

  /**
   * Number of channels per color entry: 3 (RGB) or 4 (RGBA — the alpha
   * column is per-vertex opacity, volumetric phase 4). Defaults to 3
   * when omitted. Mirrors `LoadedPointsData.colorComponents`.
   */
  colorComponents?: 3 | 4;

  /** Vertex sharpness (N,) null if not present */
  sharpness: Float32Array | null;

  /**
   * Per-vertex scalar values for colormap lookup (N,). Optional —
   * when omitted (undefined), the line shader's `USE_COLORMAP` path
   * stays inactive and `ProcessedLinesData.startScalars`/`endScalars`
   * are also undefined.
   *
   * Aligned with `LoadedPointsData.scalars?: ScalarArray` (both
   * optional + undefined absence) so consumer truthy-checks read
   * uniformly across node types.
   *
   * Dtype aligned with `ScalarArray` (Float32/Float16/Uint8) so
   * Uint8 loaders can keep native dtype through the accumulator and
   * pay a single widening at projection time rather than 4× memory up
   * front.
   */
  scalars?: ScalarArray;

  /** Number of segments loaded */
  segmentCount: number;

  /** Number of vertices loaded */
  vertexCount: number;

  /** Dimensionality for interpreting vertices array */
  ndim: number;

  /**
   * The ascending, pairwise-disjoint ON-DISK VERTEX ranges that the loaded
   * per-vertex arrays — `positions`, `widths`, `colors`, `sharpness`,
   * `scalars` — concatenate, FLATTENED as
   * `[start0, end0, start1, end1, …]`: each range is a half-open
   * `[start, end)` pair, so the array's length is always even and
   * `length / 2` is the range count. This is index space **A** in the lines
   * picking chain (see `rendering/picking/picking-system/element-id-map.ts`):
   * line labels are PER-VERTEX and their CSR is keyed by the on-disk *sorted*
   * vertex row, so these ranges are the second half of the slot → on-disk map
   * picking resolves labels through (`data/loaders/element-ids.ts`).
   *
   * A FLAT `Uint32Array` rather than an object array precisely because this
   * payload is cached: `data/loaders/progressive/slice-cache-helper.ts` bills
   * only own typed-array properties, so an object array would be retained by
   * the LRU at a billed 0 bytes (a fragmented labelled slice — the loader logs a
   * fragmentation "efficiency" diagnostic, so high range counts are expected —
   * could hold roughly twice the bytes the budget believes), and
   * `cloneLodSnapshot` would pass the very same mutable objects into the stored
   * snapshot by reference. A typed array is measured and deep-copied for free,
   * and costs ~6x fewer retained bytes per range.
   *
   * Note this is a VERTEX space, not the segment space `SegmentRange` names:
   * `LinesSpatialIndexLoader` queries visible SEGMENT ranges, then derives the
   * vertex ranges those segments reference and loads only those.
   *
   * Published ONLY when the node declares a per-element string/image CSR
   * (`has_labels` / `has_image_labels` / `has_keys`): nothing else reads the
   * map, and the field would otherwise ride along in every SliceCache snapshot
   * for free.
   * Absent ⇒ no map is composed and picking falls back to the raw slot.
   */
  vertexRangeBounds?: Uint32Array;
}

/**
 * Processed lines data ready for GPU rendering.
 *
 * After nD clipping and 3D projection:
 * - Positions are in 3D display space
 * - Per-segment attributes ready for instanced rendering
 * - Clipped endpoints have interpolated attributes
 */
export interface LinesProjectionBounds {
  /** AABB min corner over start+end positions [x, y, z]. */
  min: [number, number, number];
  /** AABB max corner over start+end positions [x, y, z]. */
  max: [number, number, number];
  /**
   * Max finite half-width over start/end widths — the conservative
   * per-segment footprint used to expand the cull box (matches
   * `computeLineBounds`' expansion semantics exactly).
   */
  maxWidth: number;
}

export interface ProcessedLinesData {
  /** Segment start positions in 3D display space (M * 3) */
  startPositions: Float32Array;

  /** Segment end positions in 3D display space (M * 3) */
  endPositions: Float32Array;

  /** Start vertex colors RGB (M * 3), interpolated if clipped */
  startColors: Float32Array;

  /** End vertex colors RGB (M * 3), interpolated if clipped */
  endColors: Float32Array;

  /** Start vertex widths (M,), interpolated if clipped */
  startWidths: Float32Array;

  /** End vertex widths (M,), interpolated if clipped */
  endWidths: Float32Array;

  /** Start vertex sharpness (M,), interpolated if clipped */
  startSharpness: Float32Array;

  /** End vertex sharpness (M,), interpolated if clipped */
  endSharpness: Float32Array;

  /** 3D segment lengths (M,), used for cap factor calculation */
  segmentLengths: Float32Array;

  /**
   * Per-endpoint joint code for the start endpoint (M,): how the line shader
   * should treat it, and — at an ordinary two-segment joint — which segment it
   * joins. `0` free polyline end (keep the soft cap); `-1` slice-clipped
   * (suppress the cap; no neighbour will arrive); `-2` degree->=3 hub (keep the
   * cap); `+(slot + 1)` joins visible segment `slot` at that segment's START;
   * `-(slot + 3)` joins it at that segment's END.
   *
   * `slot` indexes the VISIBLE stream, which is exactly a line-texture storage
   * slot, so it needs no adjustment when the depth-sort worker permutes draw
   * order. Purely topological: the bend angle is measured in screen space by
   * the vertex stage, per frame, so it tracks the camera.
   *
   * Produced by `compute_joint_codes` — see
   * `wasm/rust/src/lines_clipping.rs` for the derivation.
   */
  startJointCode: Float32Array;

  /** Same for the end endpoint (M,). */
  endJointCode: Float32Array;

  /**
   * Start-vertex scalar values (M,), interpolated if clipped, optional.
   *
   * present when the source `LoadedLinesData.scalars` is non-null.
   * Written into texel5.x of the line texture and used by the line
   * vertex shader under `USE_COLORMAP` (presence rides the
   * `userData.hasScalars` stamp — the fixed 6-texel layout always has
   * the slot).
   */
  startScalars?: Float32Array;

  /**
   * End-vertex scalar values (M,), interpolated if clipped, optional.
   * Written into texel5.y.
   */
  endScalars?: Float32Array;

  /**
   * Start-vertex opacity alphas (M,), interpolated if clipped,
   * optional. Present when the source colors carry an RGBA alpha
   * column (`LoadedLinesData.colorComponents === 4`); normalized to
   * [0, 1] float regardless of the source color dtype. Written into
   * texel5.z; presence rides the `userData.hasElementAlpha` stamp
   * (gates only the volumetric w(a) optical-depth map).
   */
  startAlphas?: Float32Array;

  /**
   * End-vertex opacity alphas (M,), interpolated if clipped, optional.
   * Written into texel5.w.
   */
  endAlphas?: Float32Array;

  /** Number of visible segments after clipping */
  segmentCount: number;

  /**
   * Fused-scan cull metadata from projection (AABB over start+end
   * positions + max finite width — see {@link LinesProjectionBounds}).
   * Mirrors the Points `metadata.bounds` / gsplats `bounds` pattern:
   * when present, `computeLineBounds` skips its O(N) per-segment scan;
   * absent (zero visible segments, or a producer that didn't scan) ⇒
   * scan fallback. Plain scalars so the object survives the worker
   * structured clone.
   */
  bounds?: LinesProjectionBounds;

  /**
   * Slot → ON-DISK element index map for per-element label lookups (M,),
   * composed at projection time by `data-processor-lines.ts`.
   *
   * Indexed by the VISIBLE SEGMENT slot the lines pick shader reports (the
   * line-texture texel row); the value is the on-disk *sorted VERTEX* row of
   * that segment's **START** vertex, because line labels are per-vertex and
   * their CSR is keyed by that row. A segment has two endpoints and the pick id
   * is a `flat` vertex-stage varying, so exactly one of them can be reported —
   * see the convention note in `data-processor-lines.ts`.
   *
   * Absent ⇒ identity / unavailable: picking falls back to the raw slot. That
   * happens for a node with no label CSR (no `vertexRangeBounds` published), and
   * whenever the composition inputs were inconsistent (fail closed). The commit
   * stamps it onto the MESH (`types/committed-data::setElementIdMap`), never
   * onto the loaded payload — that payload may be a SliceCache-owned snapshot.
   */
  elementIds?: Uint32Array;
}

// ============================================================================
// nD Clipping Types
// ============================================================================

/**
 * Result of clipping a segment to the current nD slice.
 *
 * Segments may:
 * - Be fully visible (both endpoints in slice)
 * - Be partially visible (clipped to slice boundary)
 * - Be invisible (both endpoints outside slice, same side)
 */
export interface ClippedSegment {
  /** Clipped start position in 3D display space */
  p1: number[];

  /** Clipped end position in 3D display space */
  p2: number[];

  /** Parameter t at clipped start (0-1, for attribute interpolation) */
  t1: number;

  /** Parameter t at clipped end (0-1) */
  t2: number;

  /** Whether segment should be rendered */
  visible: boolean;
}

// ============================================================================
// Scene Integration Types
// ============================================================================

import type { UpdateSession } from '../profiling/update-profiler';

/**
 * Data loader interface for Lines nodes.
 *
 * Mirrors the DataLoader interface from points but specialized for lines.
 */
export interface LinesDataLoader {
  /** Load lines data for the given view state */
  loadLines(viewState: LinesViewState, session?: UpdateSession): Promise<LoadedLinesData>;

  /** Update existing data for a new view state (see DataLoader.updateView for `signal`). */
  updateView(
    viewState: LinesViewState,
    session?: UpdateSession,
    signal?: AbortSignal,
    residencyAllowanceBytes?: number
  ): Promise<LoadedLinesData>;

  /** Clean up resources */
  dispose(): void;

  /**
   * LoaderMonitor surface (optional, for the data-loading-monitor UI).
   * Mirrors the surface that `points-spatial-index-loader.ts` exposes —
   * implementations that don't track metrics may omit these methods.
   */
  addEventListener?(listener: MonitorEventListener): void;
  removeEventListener?(listener: MonitorEventListener): void;
  getMetrics?(): LoaderMetrics;
  getActiveQueries?(): QueryInfo[];
}

/**
 * View state for lines loading.
 *
 * Identical shape to the points + gsplats ViewState — kept as a named
 * alias for documentation (a function signature reading
 * `viewState: LinesViewState` is more self-documenting than the
 * generic `ViewState`).
 */
export type LinesViewState = ViewState;

/**
 * User data attached to THREE.Mesh for Lines in scene.
 *
 * Lines use THREE.Mesh with InstancedBufferGeometry (not InstancedMesh)
 * to avoid exceeding WebGL's 16 attribute location limit.
 *
 * Enables runtime type checking and provides access to loader/metadata.
 * Note: Dimension info is NOT stored here - it's only at the Scene level.
 */
export interface LinesUserData {
  /** Node type identifier for runtime type checking */
  nodeType: 'lines';

  /** Data loader instance */
  loader: LinesDataLoader;

  /** Zarr group attributes */
  attrs: LinesMetadata;

  /** Maximum line width (for bounding box expansion) */
  maxWidth: number;

  /** Currently visible segment count after nD clipping (updated on view change) */
  visibleSegmentCount?: number;

  /** Elements requested by the latest successful commit, before the GPU capacity clamp. */
  requestedElementCount?: number;

  /** Elements omitted by the latest successful commit's GPU capacity clamp. */
  droppedElementCount?: number;

  /**
   * True when the GPU interleaved buffer holds the exact data of the last
   * commit — i.e. the append fast path may skip re-uploading the prefix.
   * Cleared to ``false`` by ``NodeFactory.rebuildAfterContextRestore`` after
   * a WebGL context loss (the CPU mirror survives but the GPU buffers are
   * gone), which forces the next commit to a full rewrite. Re-enabled by
   * every full commit. Mirrors ``GSplatsUserData.gpuPrefixIntact``.
   */
  gpuPrefixIntact?: boolean;

  /**
   * The ``SceneLoader`` view-update version this mesh's committed geometry was
   * loaded for. Written at commit (``commit-lines-geometry.ts`` via
   * ``stampLoadedViewVersion``) and read by the LOD registry to tell whether a
   * level is *fresh for the current view (slice/displayDims)* vs merely
   * committed — a re-slice overwrites geometry in place without changing
   * readiness. ``undefined`` ⇒ never committed ⇒ treated as stale.
   */
  loadedViewVersion?: number;

  /** Pick ID assigned by PickingSystem for GPU picking (undefined if picking disabled) */
  pickId?: number;
}

// ============================================================================
// Type Guards
// ============================================================================

/**
 * Check if metadata is for a Lines node.
 *
 * @param attrs - Unknown attributes object
 * @returns True if attrs is LinesMetadata
 */
export function isLinesMetadata(attrs: unknown): attrs is LinesMetadata {
  return (
    typeof attrs === 'object' &&
    attrs !== null &&
    (attrs as Record<string, unknown>).type === 'lines'
  );
}

/**
 * Check if userData indicates a Lines object.
 *
 * @param userData - THREE.Object3D userData
 * @returns True if userData is LinesUserData
 */
export function isLinesUserData(userData: unknown): userData is LinesUserData {
  return (
    typeof userData === 'object' &&
    userData !== null &&
    (userData as Record<string, unknown>).nodeType === 'lines'
  );
}

/**
 * Check if a line type is valid.
 *
 * @param lineType - String to check
 * @returns True if lineType is a valid LineType
 */
export function isValidLineType(lineType: unknown): lineType is LineType {
  return typeof lineType === 'string' && (LINE_TYPES as readonly string[]).includes(lineType);
}
